src/anyio/_backends/_asyncio.py

```diff
@@ -428,10 +428,12 @@ def _uncancel(self, cancelled_exc: CancelledError) -> bool:
             self._cancel_calls = 0
             return True
 
-        # Uncancel all AnyIO cancellations
+        # Undo all cancellations done by this scope
         if sys.version_info >= (3, 11):
-            for i in range(self._cancel_calls):
-                self._host_task.uncancel()
+            while self._cancel_calls:
+                self._cancel_calls -= 1
+                if not self._host_task.uncancel() and not self._cancel_calls:
```

Comment:
I think cancel scopes need to store the current task.cancelling() value and compare to that rather than 0 like asyncio.timeout does in 1.11.4

Reply:
```python
try:
    asyncio.current_task.cancel()
    await anyio.sleep(1)
except CancelledError:
        try:
            with CancelScope() as scope:
                scope.cancel()
                try:
                    await anyio.sleep(0)
                except asyncio.CancelledError:
                    raise asyncio.CancelledError
        except asyncio.CancelledError:
            pytest.fail("Should have swallowed the CancelledError")
```

Reply:
https://github.com/python/cpython/pull/102815/files#diff-0a6da8818df8b29c3c649167aa37c906eb27283c5fe1cacb40e36a6ce0458e7eR87